## Individual assignment

**Humanity has finally reached Mars. There are 4 colonies, and the humans are more or less adapting to life there**

The 4 cities are called `Yharnam`, `Marsdrid`, `Ironforge` and `Churroland`. Each city has its own `population`.
```python
population = {
    'Yharnam': 1000,
    'Marsdrid': 2000,
    'Ironforge': 3000,
    'Churroland': 4000,
}
``` 

The cities are all connected to each other, with different `distances` between them (see attached PDF). There are also important points for life, like the three water sources, or wells. There's also a volcano that the Mars Geological Survey is monitoring. All these landmarks are defined by their coordinates, which are given in the `landmarks` dictionary. 
    
```python
# coordinates in kilometers
landmarks = {
    'Yharnam': (0, 0),
    'Marsdrid': (0, 150),
    'Ironforge': (100, 50),
    'Churroland': (150, 100),
    'Well 1': (0, 50),
    'Well 2': (0, 100),
    'Well 3': (100, 100),
    'Volcano': (50, 100),
}
```

Rules:
* You can use any of the tools we have seen in class, nothing else.
* You can't share your code with anyone else.
* 1 point per correctly solved exercise.
* No points deducted from incorrect answers.
* **SUBMISSION:**
  * ONE (1) JUPYTER NOTEBOOK, based on this one, with your name as the filename:
    * `individual_assignment_daniel_garcia_hernandez.ipynb`
  * No submission will be accepted after the deadline.
  * No submission will be accepted if the description of the exercises is changed
  * No submission will be accepted if the order of the exervises is changed

In [1]:
population = {
    'Yharnam': 1000,
    'Marsdrid': 2000,
    'Ironforge': 3000,
    'Churroland': 4000,
}

landmarks = {
    'Yharnam': (0, 0),
    'Marsdrid': (0, 150),
    'Ironforge': (100, 50),
    'Churroland': (150, 100),
    'Well 1': (0, 50),
    'Well 2': (0, 100),
    'Well 3': (100, 100),
    'Volcano': (50, 100),
}

### Exercise 1

Knowing that we can calculate the distance between two points in 2-dimensional space using the formula:
$$d = \sqrt{(x_2 - x_1)^2 + (y_2 - y_1)^2}$$
write a function that takes two points as input and returns the distance between them.

Each point should be a tuple of two numbers `(x, y)` representing the two coordinates --horizontal and vertical-- of the point.

```python
def distance(point1, point2):
    # Your code here
    return distance
```

In [2]:
import math

In [276]:
def distance(point1, point2):
    x1, y1 = point1
    x2, y2 = point2
    distance = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)
    return distance

In [307]:
point1 = (200,50)
point2 = (150, 100)
result_formula = distance(point1, point2)
print(result_formula)

70.71067811865476


### Exercise 2

Now that we know how to calculate the distance between two points, we can calculate the distance between two landmarks. 

Write a function that takes the followning arguments:
* `landmarks`: a dictionary with the coordinates of the landmarks
* `landmarks_list`: a list of the landmarks we want to calculate the distance between. It should be a list of 2 strings, where each string is the name of a landmark.

and returns the distance between them, using the function you wrote in the previous exercise.

```python
def distance_between_landmarks(landmarks, landmarks_list):
    # Your code here using the distance function from exercise 1
    return distance
```

In [311]:
def distance_between_landmarks(landmarks, landmarks_list):
    point1 = landmarks.get(landmarks_list[0])
    point2 = landmarks.get(landmarks_list[1])
    return distance(point1, point2)

In [312]:
landmarks_list = ['Marsdrid', 'Yharnam']
result_landmark = distance_between_landmarks(landmarks, landmarks_list)
print(result_landmark)

150.0


### Exercise 3

Let's analyze which city has a better access to water. To do so, we need to calculate the distance between each city and each well, and then average the distances.

We can do this by using a double `for` loop:

```python
dict_of_avg_distances = {}

for city in cities:

    distances_to_wells = []

    for well in wells:
        
        # Calculate distance between city and well
        # save the distance in a list with append

    # sum the 3 distances in the list and divide by 3 to calculate the average
    # add the average distance to a dictionary where the key is the city and the value is the average distance to water.
```


In [278]:
def analyze_water_access(landmarks):
    dict_of_avg_distances = {}

    cities = [name for name in landmarks if name not in ['Well 1', 'Well 2', 'Well 3', 'Volcano']]
    wells = [landmarks[name] for name in landmarks if name.startswith('Well')]

    for city in cities:
        distances_to_wells = []

        for well in wells:
            distance_to_well = distance(landmarks[city], well)
            distances_to_wells.append(distance_to_well)

        avg_distance = sum(distances_to_wells) / len(distances_to_wells)
        dict_of_avg_distances[city] = avg_distance

    return dict_of_avg_distances


In [313]:
avg_distances_to_well = analyze_water_access(landmarks)
print(avg_distances_to_well)

wells

{'Yharnam': 97.14045207910317, 'Marsdrid': 87.2677996249965, 'Ironforge': 87.2677996249965, 'Churroland': 119.37129433613966}


### Exercise 4

Which city or cities have the best access to water? Which city or cities has the worst access to water?

Hint: you can use the `min` and `max` functions to find the minimum and maximum values in a dictionary, passing the dictionary as an argument, and using the `key` argument to specify which value to use to find the minimum or maximum.

```python
dummy_dict = {
    'city1': 10,
    'city2': 20,
    'city3': 30,
}

min_value = min(dummy_dict, key=dummy_dict.get)
max_value = max(dummy_dict, key=dummy_dict.get)
```
Keep in mind that these functions will return the first key found with the minimum or maximum value, so if there are repeated values, it might not be the one you want.

In [324]:
#For this calculation, the best and worst cities access to well is by using the averages calculated before. 

avg_distances_wells = {
    'Yharnam': 97.14,
    'Marsdrid': 87.26,
    'Ironforge': 87.26,
    'Churroland': 119.37
}

best_distance = min(avg_distances_wells.values())
worst_distance = max(avg_distances_wells.values())

best_cities = [city for city, distance in avg_distances_wells.items() if distance == best_distance]
worst_cities = [city for city, distance in avg_distances_wells.items() if distance == worst_distance]

print("Cities with the best access to water:", best_cities)
print("Cities with the worst access to water:", worst_cities)

#We can see that 2 cities have the same average distance to the Wells. 

Cities with the best access to water: ['Marsdrid', 'Ironforge']
Cities with the worst access to water: ['Churroland']


### Exercise 5

Now that we know which city has the best access to water, create a dictionary with the cities as keys and their distance to the volcano as values

In [282]:
volcano_distances = landmarks['Volcano']

city_distances_to_volcano = {}

for city, coordinates in landmarks.items():
    if city != 'Volcano' and not city.startswith('Well'):
        distance_to_volcano = distance(coordinates, volcano_distances)
        city_distances_to_volcano[city] = distance_to_volcano

print(city_distances_to_volcano)

{'Yharnam': 111.80339887498948, 'Marsdrid': 70.71067811865476, 'Ironforge': 70.71067811865476, 'Churroland': 100.0}


### Exercise 6

The Geological Survey found that in case of an eruption, the explosion will have a radius between 75 and 110 km.
* In the best case scenario, the explosion will be 75 km away from the volcano.
* In the worst case scenario, the explosion will be 110 km away from the volcano.

Knowing this, which city or cities are at risk of being destroyed by the volcano in the best case scenario? Which city or cities are at risk of being destroyed by the volcano in the worst case scenario?

Write a function that takes in the `landmarks` and the explosion radii in each scenario, and returns which cities are at rist of being destroyed.

```python
def cities_at_risk(landmarks, [best_case_radius, worst_case_radius]):
    # Your code here
    return cities_at_risk_best_case, cities_at_risk_worst_case
```

In [284]:
def cities_at_risk(landmarks, explosion_radius):
    best_case_radius, worst_case_radius = explosion_radius
    
    cities_at_risk_best_case = [city for city, distance in landmarks.items() if distance <= best_case_radius]
    cities_at_risk_worst_case = [city for city, distance in landmarks.items() if distance <= worst_case_radius]
    
    return cities_at_risk_best_case, cities_at_risk_worst_case

explosion_radius = [75, 110]

cities_at_risk_best_case, cities_at_risk_worst_case = cities_at_risk(city_distances_to_volcano, explosion_radius)

print("Cities at risk in the best case scenario:", cities_at_risk_best_case)
print("Cities at risk in the worst case scenario:", cities_at_risk_worst_case)


Cities at risk in the best case scenario: ['Marsdrid', 'Ironforge']
Cities at risk in the worst case scenario: ['Marsdrid', 'Ironforge', 'Churroland']


### Exercise 7

Our worst fears have come true. The volcano has increased its activity and our geologist have estimated the radius of explosion: 90 km.

We need to evacuate the cities at risk. How many people will we have to evacuate?

Adapt the previous function to return the number of people at risk, passing the `population` dictionary as an argument, as well as `landmarks` and the explosion radius.

```python
def population_to_evacuate(landmarks, population, explosion_radius):
    # Your code here
    return people_at_risk
```

In [285]:
def population_to_evacuate(landmarks, population, explosion_updated):
    cities_at_risk = [city for city, distance in landmarks.items() if distance <= explosion_updated]
    people_at_risk = sum(population.get(city, 0) for city in cities_at_risk)
    return people_at_risk

explosion_updated = 90

people_to_evacuate = population_to_evacuate(city_distances_to_volcano, population, explosion_updated)

print("Number of people to evacuate:", people_to_evacuate)

Number of people to evacuate: 5000


### Exercise 8

Which cities are the ones we need to evacuate? Which cities are the ones we can keep?

Reuse the previous function to return a dictionary with `keep` and `evacuate` as keys, and the cities as values.

```python
dict_status = {
    'keep': [city1, city2, city3],
    'evacuate': [city4, city5, city6],
}
```

```python
def status_of_cities(landmarks, explosion_radius):
    # Your code here
    return dict_status
```

In [287]:
def status_of_cities(landmarks, explosion_updated):
    cities_at_risk = [city for city, distance in landmarks.items() if distance <= explosion_updated]
    all_cities = list(landmarks.keys())

    dict_status = {
        'keep': [city for city in all_cities if city not in cities_at_risk],
        'evacuate': cities_at_risk
    }

    return dict_status


explosion_updated = 90
dict_status = status_of_cities(city_distances_to_volcano, explosion_updated)

cities_to_keep = dict_status['keep']
cities_to_evacuate = dict_status['evacuate']

print("Cities to Keep:", cities_to_keep)
print("Cities to Evacuate:", cities_to_evacuate)


Cities to Keep: ['Yharnam', 'Churroland']
Cities to Evacuate: ['Marsdrid', 'Ironforge']


### Exercise 9: open-ended

What's the best way to evacuate the cities? Let's think of the difficulty for the evacuation as the product of the population to be evacuated and the distance to which they're being evacuated.

```python
difficulty = population_to_evacuate * distance_to_safety
```

We can evacuate all the population to the same city, or we can split the population between the cities, according to the distance between their original city and the city they're being evacuated to.

Use code to back your decision.

#### Open-Ended Solution

The steps necessary to solve this question are the following:
    1) The cities to evacuate are determined in Exercise 8, which takes into account the new explosion radius of the volcano, which is 90km. The cities to evacuate are "Marsdrid and Ironforge", and the cities that could be used as evacuation "safety" cities are "Churroland and Yharnam". 
    2) Identify the Scenario's available to evacuate the cities, based on the formula given before called "difficulty", which determines the easiest possible scenario for evacuating the city. 
    3) Take the decision of which scenario is best/worse, backed with the coding.
    
The two possible scenario's are the following: 
        Scenario 1: Evacuating all the cities in danger to the same evacuation city. In this scenario there can be 2 options, all to Yharnam or all to Churroland.
               (Marsdrid + Ironforge --> Churroland) or 
               (Marsdrid + Ironforge --> Yharnam)      
        Scenario 2: Evacuating the danger cities to the lowest difficulty city. In this scenario, the evaluation is separately of each city. Once you evaluate the lowest score, choose the best possible evacuation city for the danger city. The total difficulty of Scenario 2 is the sum of both evacuations lowest possibilities. 
                 (Marsdrid --> Churroland) or (Marsdrid --> Yharnam) 
                and (Ironforge --> Churroland) or (Ironforge --> Yharnam) 

In [251]:
#Scenario 1 
# The distance to safety is calculated using the function in exercise 1 or 2 and inserted here. 
evacuate_to_churroland = (population['Marsdrid'] * 158.13) +  (population['Ironforge'] * 70.71) 
evacuate_to_yharnam = (population['Marsdrid'] * 150) +  (population['Ironforge'] * 111.80) 

min_difficulty = min(evacuate_to_yharnam, evacuate_to_churroland)

if min_difficulty == evacuate_to_yharnam:
    best_evacuation_city = 'Yharnam'
else:
    best_evacuation_city = 'Churroland'
    
#Printing the results to know which option is better in this case scenario
print("Difficulty of evacuating to Yharnam:", evacuate_to_yharnam)
print("Difficulty of evacuating to Churroland:", evacuate_to_churroland)
print("Best city for evacuation:", best_evacuation_city,", Difficulty:", min_difficulty)

Difficulty of evacuating to Yharnam: 635400.0
Difficulty of evacuating to Churroland: 528390.0
Best city for evacuation: Churroland , Difficulty: 528390.0


In [244]:
#Scenario 2: Evacuate danger city to the city with the lowest difficulty

In [245]:
#Marsdrid to Churroland or to Yharnam 
# The distance to safety is calculated using the function in exercise 1 or 2 and inserted here. 
evacuate_marsdrid_to_churroland = (population['Marsdrid'] * 158.13)
evacuate_marsdrid_to_yharnam = (population['Marsdrid'] * 150)

min_difficulty_marsdrid = min(evacuate_marsdrid_to_churroland,evacuate_marsdrid_to_yharnam)

if min_difficulty_marsdrid == evacuate_marsdrid_to_yharnam:
    best_evacuation_city_marsdrid = 'Yharnam'
else:
    best_evacuation_city_marsdrid = 'Churroland'

print("Difficulty of evacuating Marsdrid to Yharnam:", evacuate_marsdrid_to_yharnam)
print("Difficulty of evacuating Marsdrid to Churroland:",evacuate_marsdrid_to_churroland)
print("Best city for evacuation of Marsdrid:", best_evacuation_city_marsdrid,", Difficulty:", min_difficulty_marsdrid)


Difficulty of evacuating Marsdrid to Yharnam: 300000
Difficulty of evacuating Marsdrid to Churroland: 316260.0
Best city for evacuation of Marsdrid: Yharnam , Difficulty: 300000


In [ ]:
#Ironforge to Churroland or to Yharnam 
#The distance to safety is calculated using the function in exercise 1 or 2 and inserted here. 
evacuate_ironforge_to_churroland = (population['Ironforge'] * 70.71)
evacuate_irongforge_to_yharnam = (population['Ironforge'] * 111.80) 

min_difficulty_ironforge = min(evacuate_ironforge_to_churroland,evacuate_irongforge_to_yharnam)

if min_difficulty_ironforge == evacuate_irongforge_to_yharnam:
    best_evacuation_city_ironforge = 'Yharnam'
else:
    best_evacuation_city_ironforge = 'Churroland'

print("Difficulty of evacuating Ironforge to Yharnam:", evacuate_irongforge_to_yharnam)
print("Difficulty of evacuating Ironforge to Churroland:",evacuate_ironforge_to_churroland)
print("Best city for evacuation of Ironforge:", best_evacuation_city_ironforge,", Difficulty:", min_difficulty_ironforge)


Difficulty of evacuating Ironforge to Yharnam: 335400.0
Difficulty of evacuating Ironforge to Churroland: 212129.99999999997
Best city for evacuation of Ironforge: Churroland , Difficulty: 212129.99999999997


In [253]:
#Total Difficulty of Scenario 1: 
print ("Minimum Possible Difficulty Scenario 1:", min_difficulty, "which is evacuating Marsdrid and Ironforge to:",best_evacuation_city)

#Total Difficulty of Scenario 2:
print ("Minimum Possible Difficulty Scenario 2:", min_difficulty_ironforge + min_difficulty_marsdrid , "which is evacuating Marsdrid to", best_evacuation_city_marsdrid, ", and Ironforge to",best_evacuation_city_ironforge) 

Minimum Possible Difficulty Scenario 1: 528390.0 which is evacuating Marsdrid and Ironforge to: Churroland
Minimum Possible Difficulty Scenario 2: 512130.0 which is evacuating Marsdrid to Yharnam , and Ironforge to Churroland


#### Decision of Evacuation 

The best possible evacuation plan derived from the scenarios evaluated, leaving a side if any of the cities has a constraint of allowing an extra percentage of population. 

Scenario 1: The best possible solution of this scenario is to evacuate Marsdrid and Ironforge to Churroland, which derives with the lowest possible difficulty at 528390. 

Scenario 2: The evaluation is separate of each city. The the difficulty of the best option of each is summed up. 
Marsdrid: The best option for Marsdrid is to be evacuated to Yharnam, with a difficulty of 300000. 
Ironforge: The best option for Ironforge is to be evacuated to Churroland, with a difficulty of 212129.99

To conclude, the best possible way to evacuate the cities in danger is given in Scenario 2. It sums a difficulty of 512130 compared to Scenario 1, 528390. Adding to the sum of the difficulty we could also add the potential collapse of the city in the case of scenario 1, which is managing an extra 5000 people in their city.